clash-lib/src/Clash/Normalize/Transformations.hs

```diff
@@ -494,150 +490,150 @@ inlineNonRep _ e = return e
-                          ,"GHC.Err.undefined"] ->
-                let e' = mkApps (mkTicks (Prim ty') ticks)
-                                [repTy,Right ty,msgOrCallStack]
-                in  changed e'
-            (Prim ty',[_],ticks)
-              | primName ty' `elem` [ "Clash.Transformations.undefined"
-                          , "Clash.GHC.Evaluator.undefined"
-                          , "EmptyCase"] ->
-                let e' = mkApps (mkTicks (Prim ty') ticks) [Right ty]
-                in changed e'
-            _ -> do
-              let subjTy = termType tcm subj
-              tran <- Lens.view typeTranslator
-              case (`evalState` HashMapS.empty) (coreTypeToHWType tran reprs tcm subjTy) of
-                Right (FilteredHWType (Void (Just hty)) _areVoids)
-                  | hty `elem` [BitVector 0, Unsigned 0, Signed 0, Index 1]
-                  -> caseCon ctx' (Case (Literal (IntegerLiteral 0)) ty alts)
-                _ -> do
-                  let ret = caseOneAlt e
-                  if lvl > DebugNone then do
-                    let subjIsConst = isConstant subj
-                    traceIf (lvl > DebugNone && subjIsConst) ("Irreducible constant as case subject: " ++ showPpr subj ++ "\nCan be reduced to: " ++ showPpr subj') ret
-                  else
-                    ret
-
-caseCon ctx e@(Case subj ty alts) = do
-  reprs <- Lens.view customReprs
-  tcm <- Lens.view tcCache
-  let subjTy = termType tcm subj
-  tran <- Lens.view typeTranslator
-  case (`evalState` HashMapS.empty) (coreTypeToHWType tran reprs tcm subjTy) of
-    Right (FilteredHWType (Void (Just hty)) _areVoids)
-      | hty `elem` [BitVector 0, Unsigned 0, Signed 0, Index 1]
-      -> caseCon ctx (Case (Literal (IntegerLiteral 0)) ty alts)
-    _ -> caseOneAlt e
+                                ,"GHC.Err.undefined"] ->
+        let e1 = mkApps (mkTicks (Prim pInfo) ticks)
+                        [repTy,Right ty,msgOrCallStack]
+        in  changed e1
+      -- WHNF of subject is _|_, in the form of our internal _|_-values: that
+      -- means the entire case-expression is _|_
+      (Prim pInfo,[_],ticks)
+        | primName pInfo `elem` [ "Clash.Transformations.undefined"
+                                , "Clash.GHC.Evaluator.undefined"
+                                , "EmptyCase"] ->
+        let e1 = mkApps (mkTicks (Prim pInfo) ticks) [Right ty]
+        in changed e1
+      -- WHNF of subject is non of the above, so either a variable reference,
+      -- or a primitive for which the evaluator doesn't have any evaluation
+      -- rules.
+      _ -> do
+        tcm <- Lens.view tcCache
+        let subjTy = termType tcm subj
+        tran <- Lens.view typeTranslator
+        reprs <- Lens.view customReprs
+        case (`evalState` HashMapS.empty) (coreTypeToHWType tran reprs tcm subjTy) of
+          Right (FilteredHWType (Void (Just hty)) _areVoids)
+            | hty `elem` [BitVector 0, Unsigned 0, Signed 0, Index 1]
```

Comment:
Couldn't we make this a bit more general and use `typeSize`?

Reply:
No... we really only want to do it for those types; so we'd need to extract the `hty` out of the `Void`.